File: apps/sdk-typescript/src/http.ts
/**
 * Lightweight fetch wrapper used by every service client.
 *
 * Responsibilities:
 *   - URL building (base + path + querystring)
 *   - Auth header injection (`Authorization: Bearer ...` if `authToken` set)
 *   - JSON encoding / decoding
 *   - Envelope unwrapping (`{ ok: true, data }` → data, `{ ok: false, error }` → throw)
 *   - Timeout via `AbortController`
 *   - Retry with exponential backoff on 5xx and fetch-level failures
 *
 * The wrapper is intentionally small and dependency-free. Tests inject a
 * stubbed `fetch` (typically MSW's interceptor) via the `PraxisClient`
 * constructor so this module is exercised in lockstep with the service
 * clients.
 */

import { isErrorEnvelope, isOkEnvelope } from './envelope.js';
import { PraxisApiError, PraxisNetworkError } from './errors.js';

export type FetchLike = (input: string | URL | Request, init?: RequestInit) => Promise<Response>;

export interface RetryConfig {
  /** Maximum extra attempts after the first try. `2` => up to 3 total tries. */
  readonly count: number;
  /** Initial backoff in ms; subsequent attempts double the delay. */
  readonly backoffMs: number;
}

export interface HttpClientOptions {
  readonly fetch: FetchLike;
  readonly timeoutMs: number;
  readonly retries: RetryConfig;
  readonly authToken?: string;
  /**
   * When set, the client pauses for the requested ms between retries. Tests
   * inject a stub that returns immediately so retry semantics can be verified
   * without real timers.
   */
  readonly sleep?: (ms: number) => Promise<void>;
}

export interface RequestParams {
  readonly method: 'GET' | 'POST' | 'PATCH' | 'DELETE' | 'PUT';
  readonly baseUrl: string;
  readonly path: string;
  readonly query?: Readonly<Record<string, string | number | boolean | undefined>>;
  readonly body?: unknown;
  /**
   * When `true`, treat the response body as `void` (no envelope, no JSON
   * decode). Used for 204-No-Content endpoints like `DELETE /alerts/:id`.
   */
  readonly expectNoBody?: boolean;
}

const defaultSleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Pure helper — joins `base` and `path` and appends a querystring built from
 * `query` (skipping `undefined` values). Exported for tests.
 */
export const buildUrl = (
  base: string,
  path: string,
  query?: Readonly<Record<string, string | number | boolean | undefined>>,
): string => {
  const trimmedBase = base.replace(/\/+$/, '');
  const normalizedPath = path.startsWith('/') ? path : `/${path}`;
  let url = `${trimmedBase}${normalizedPath}`;
  if (query) {
    const params = new URLSearchParams();
    for (const [key, value] of Object.entries(query)) {
      if (value === undefined) {
        continue;
      }
      params.append(key, String(value));
    }
    const qs = params.toString();
    if (qs.length > 0) {
      url += `?${qs}`;
    }
  }
  return url;
};

const isRetriableStatus = (status: number): boolean => status >= 500 && status <= 599;

/**
 * Run a single HTTP attempt, including timeout. Resolves with the parsed body
 * (already envelope-checked) on 2xx, or throws `PraxisApiError` on a 4xx/5xx
 * with a parseable error envelope. Throws `PraxisNetworkError` on transport
 * failures; the caller decides whether to retry.
 */
const runOnce = async <T>(
  options: HttpClientOptions,
  params: RequestParams,
): Promise<T | undefined> => {
  const url = buildUrl(params.baseUrl, params.path, params.query);

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), options.timeoutMs);

  const headers: Record<string, string> = {
    accept: 'application/json',
  };
  if (params.body !== undefined) {
    headers['content-type'] = 'application/json';
  }
  if (options.authToken !== undefined) {
    headers.authorization = `Bearer ${options.authToken}`;
  }

  let response: Response;
  try {
    response = await options.fetch(url, {
      method: params.method,
      headers,
      ...(params.body !== undefined ? { body: JSON.stringify(params.body) } : {}),
      signal: controller.signal,
    });
  } catch (cause) {
    if (cause instanceof Error && cause.name === 'AbortError') {
      throw new PraxisNetworkError({
        code: 'TIMEOUT',
        message: `Request timed out after ${options.timeoutMs}ms: ${params.method} ${url}`,
        cause,
      });
    }
    throw new PraxisNetworkError({
      code: 'FETCH_FAILED',
      message: `fetch failed: ${params.method} ${url}`,
      cause,
    });
  } finally {
    clearTimeout(timeout);
  }

  // 204 / explicit no-body — short-circuit.
  if (params.expectNoBody === true || response.status === 204) {
    if (!response.ok) {
      // 4xx with no body shouldn't happen from our services, but be defensive.
      throw new PraxisApiError({
        code: 'HTTP_ERROR',
        message: `HTTP ${response.status} ${response.statusText}`,
        status: response.status,
      });
    }
    return undefined;
  }

  let parsed: unknown;
  try {
    parsed = await response.json();
  } catch (cause) {
    throw new PraxisNetworkError({
      code: 'INVALID_JSON',
      message: `failed to parse JSON response: ${params.method} ${url}`,
      cause,
    });
  }

  if (response.ok) {
    if (!isOkEnvelope<T>(parsed)) {
      throw new PraxisNetworkError({
        code: 'INVALID_RESPONSE',
        message: `unexpected response shape (missing ok/data): ${params.method} ${url}`,
      });
    }
    return parsed.data;
  }

  if (isErrorEnvelope(parsed)) {
    throw PraxisApiError.fromBody(response.status, parsed.error);
  }

  throw new PraxisApiError({
    code: 'HTTP_ERROR',
    message: `HTTP ${response.status} ${response.statusText}`,
    status: response.status,
  });
};

/**
 * Outer driver — runs `runOnce` up to `retries.count + 1` times, retrying on
 * 5xx `PraxisApiError` and on `PraxisNetworkError` (except TIMEOUT, which
 * indicates the user budget is exhausted on this call). Backoff is
 * exponential: `backoffMs * 2^attempt`.
 */
export const request = async <T>(
  options: HttpClientOptions,
  params: RequestParams,
): Promise<T | undefined> => {
  const sleep = options.sleep ?? defaultSleep;
  const maxAttempts = options.retries.count + 1;
  let attempt = 0;
  let lastError: unknown;

  while (attempt < maxAttempts) {
    try {
      return await runOnce<T>(options, params);
    } catch (err) {
      lastError = err;

      const isLast = attempt === maxAttempts - 1;
      if (isLast) {
        break;
      }

      // Decide whether to retry. We retry on:
      //   - PraxisNetworkError with code FETCH_FAILED or INVALID_JSON / INVALID_RESPONSE
      //   - PraxisApiError with HTTP 5xx
      // We do NOT retry on:
      //   - TIMEOUT (user budget already exhausted)
      //   - 4xx (client error — replaying won't help)
      let retriable = false;
      if (err instanceof PraxisNetworkError) {
        retriable = err.code !== 'TIMEOUT';
      } else if (err instanceof PraxisApiError) {
        retriable = isRetriableStatus(err.status);
      }
      if (!retriable) {
        break;
      }

      const delay = options.retries.backoffMs * Math.pow(2, attempt);
      await sleep(delay);
      attempt += 1;
    }
  }

  throw lastError;
};

File: apps/sdk-python/src/colber_sdk/_http.py
"""Lightweight httpx wrapper used by every service client.

Responsibilities:
  - URL building (base + path + querystring with ``None`` values skipped)
  - Auth header injection (``Authorization: Bearer ...`` if ``auth_token`` set)
  - JSON encoding / decoding
  - Envelope unwrapping (``{"ok": true, "data"}`` -> data,
    ``{"ok": false, "error"}`` -> raise)
  - Timeout via httpx
  - Retry with exponential backoff on 5xx and transport failures

Mirrors ``apps/sdk-typescript/src/http.ts`` byte-for-byte in semantics.
The wrapper is deliberately small and dependency-free beyond httpx so the
SDK has minimal runtime weight.
"""

from __future__ import annotations

import json as _json
import time
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any, Literal
from urllib.parse import quote

import httpx

from .envelope import is_error_envelope, is_ok_envelope
from .errors import ColberApiError, ColberNetworkError
from .types import RetryConfig

HttpMethod = Literal["GET", "POST", "PATCH", "DELETE", "PUT"]

# Default sleep used between retries. Tests inject a no-op stub.
SleepFunc = Callable[[float], None]


@dataclass(frozen=True, slots=True)
class HttpClientOptions:
    """Bundle of options shared across every service call.

    Constructed once by :class:`ColberClient` and forwarded into every
    service-level call.
    """

    fetch: Callable[..., httpx.Response]
    timeout_s: float
    retries: RetryConfig
    auth_token: str | None = None
    sleep: SleepFunc | None = None


@dataclass(frozen=True, slots=True)
class RequestParams:
    """Per-request parameters."""

    method: HttpMethod
    base_url: str
    path: str
    query: dict[str, str | int | float | bool | None] | None = None
    body: Any = None
    expect_no_body: bool = False
    # Sentinel used to differentiate "no body" from "body was None".
    # We rely on `body=None` actually meaning no body, so callers shouldn't
    # send a literal None payload.


def _default_sleep(seconds: float) -> None:
    """Default inter-retry sleep (real time)."""
    time.sleep(seconds)


def build_url(
    base: str,
    path: str,
    query: dict[str, str | int | float | bool | None] | None = None,
) -> str:
    """Pure helper — joins ``base`` and ``path`` and appends a querystring.

    - Trailing slashes are stripped from ``base``.
    - ``path`` is normalised to start with ``/``.
    - ``None`` query values are skipped.
    - The querystring is omitted entirely when no values remain.

    Mirror of the TS ``buildUrl`` helper in ``http.ts``.
    """
    trimmed_base = base.rstrip("/")
    normalised_path = path if path.startswith("/") else f"/{path}"
    url = f"{trimmed_base}{normalised_path}"
    if query:
        parts: list[str] = []
        for key, value in query.items():
            if value is None:
                continue
            # `quote_via=quote` escapes URL-unsafe chars consistently
            # with the TS URLSearchParams behaviour.
            parts.append(f"{quote(str(key), safe='')}={quote(str(value), safe='')}")
        if parts:
            url += "?" + "&".join(parts)
    return url


def _is_retriable_status(status: int) -> bool:
    """5xx — retriable. Mirror of ``isRetriableStatus`` in TS."""
    return 500 <= status <= 599


def _run_once(
    options: HttpClientOptions,
    params: RequestParams,
) -> Any:
    """Run a single HTTP attempt, including timeout.

    Returns the parsed body (already envelope-checked) on 2xx, or raises
    :class:`ColberApiError` on a 4xx/5xx with a parseable error envelope.
    Raises :class:`ColberNetworkError` on transport failures; the caller
    decides whether to retry.
    """
    url = build_url(params.base_url, params.path, params.query)

    headers: dict[str, str] = {"accept": "application/json"}
    if params.body is not None:
        headers["content-type"] = "application/json"
    if options.auth_token is not None:
        headers["authorization"] = f"Bearer {options.auth_token}"

    request_kwargs: dict[str, Any] = {
        "method": params.method,
        "url": url,
        "headers": headers,
        "timeout": options.timeout_s,
    }
    if params.body is not None:
        request_kwargs["content"] = _json.dumps(params.body, ensure_ascii=False).encode("utf-8")

    try:
        response = options.fetch(**request_kwargs)
    except httpx.TimeoutException as cause:
        raise ColberNetworkError(
            code="TIMEOUT",
            message=f"Request timed out after {options.timeout_s}s: {params.method} {url}",
            cause=cause,
        ) from cause
    except httpx.HTTPError as cause:
        raise ColberNetworkError(
            code="FETCH_FAILED",
            message=f"fetch failed: {params.method} {url}",
            cause=cause,
        ) from cause

    # 204 / explicit no-body — short-circuit.
    if params.expect_no_body or response.status_code == 204:
        if not (200 <= response.status_code < 300):
            raise ColberApiError(
                code="HTTP_ERROR",
                message=f"HTTP {response.status_code} {response.reason_phrase}",
                status=response.status_code,
            )
        return None

    try:
        parsed: Any = response.json()
    except (ValueError, _json.JSONDecodeError) as cause:
        raise ColberNetworkError(
            code="INVALID_JSON",
            message=f"failed to parse JSON response: {params.method} {url}",
            cause=cause,
        ) from cause

    if 200 <= response.status_code < 300:
        if not is_ok_envelope(parsed):
            raise ColberNetworkError(
                code="INVALID_RESPONSE",
                message=f"unexpected response shape (missing ok/data): {params.method} {url}",
            )
        return parsed["data"]

    if is_error_envelope(parsed):
        raise ColberApiError.from_body(response.status_code, parsed["error"])

    raise ColberApiError(
        code="HTTP_ERROR",
        message=f"HTTP {response.status_code} {response.reason_phrase}",
        status=response.status_code,
    )


def request(options: HttpClientOptions, params: RequestParams) -> Any:
    """Outer driver — runs ``_run_once`` up to ``retries.count + 1`` times.

    Retries on:
      - :class:`ColberNetworkError` (except ``TIMEOUT``).
      - :class:`ColberApiError` with HTTP 5xx.

    Does NOT retry on:
      - ``TIMEOUT`` (user budget already exhausted).
      - 4xx (client error — replaying won't help).

    Backoff is exponential: ``backoff_ms * 2^attempt``.
    """
    sleep = options.sleep if options.sleep is not None else _default_sleep
    max_attempts = options.retries.count + 1
    attempt = 0
    last_error: BaseException | None = None

    while attempt < max_attempts:
        try:
            return _run_once(options, params)
        except (ColberNetworkError, ColberApiError) as err:
            last_error = err

            is_last = attempt == max_attempts - 1
            if is_last:
                break

            retriable = False
            if isinstance(err, ColberNetworkError):
                retriable = err.code != "TIMEOUT"
            elif isinstance(err, ColberApiError):
                retriable = _is_retriable_status(err.status)
            if not retriable:
                break

            delay_ms = options.retries.backoff_ms * (2**attempt)
            sleep(delay_ms / 1000.0)
            attempt += 1

    assert last_error is not None
    raise last_error


# Re-exported for typing-friendliness elsewhere in the SDK.
__all__ = [
    "HttpClientOptions",
    "HttpMethod",
    "RequestParams",
    "build_url",
    "request",
]


# `field` import is here for forward-compat — kept off `__all__`.
_ = field
